### DISCUSSION TITLE
How to understand the CHAR_LITERAL rule in java lexer file

### DISCUSSION THREAD

**Discussion by NeoZephyr** [antlr/grammars-v4#3313] [General]
```g4
CHAR_LITERAL:       '\'' (~['\\\r\n] | EscapeSequence) '\'';

fragment EscapeSequence
    : '\\' 'u005c'? [btnfr"'\\]
    | '\\' 'u005c'? ([0-3]? [0-7])? [0-7]
    | '\\' 'u'+ HexDigit HexDigit HexDigit HexDigit
    ;
```

Why '\r', '\n', '\'', '\\' is excluded in first part，and '\b', '\t', '\f', '\"' not excluded in first part?

If I change the rule to this, is it equivalent to the previous rule

```g4
CHAR_LITERAL:       '\'' (~['\\\r\n\b\f\t\"] | EscapeSequence) '\'';
```

Or change it to this

```g4
CHAR_LITERAL:       '\'' (~['\\] | EscapeSequence) '\'';
```

**Comment by msagca** [comment]
Hi @NeoZephyr 

You should refer to the Java [specification](https://docs.oracle.com/javase/specs/jls/se20/html/jls-3.html#jls-3.10.4) to make sense of this rule.

**CharacterLiteral:**
' [SingleCharacter](https://docs.oracle.com/javase/specs/jls/se20/html/jls-3.html#jls-SingleCharacter) '
' [EscapeSequence](https://docs.oracle.com/javase/specs/jls/se20/html/jls-3.html#jls-EscapeSequence) '

**SingleCharacter:**
[InputCharacter](https://docs.oracle.com/javase/specs/jls/se20/html/jls-3.html#jls-InputCharacter) but not ' or \

**InputCharacter:**
[UnicodeInputCharacter](https://docs.oracle.com/javase/specs/jls/se20/html/jls-3.html#jls-UnicodeInputCharacter) but not CR or LF

**EscapeSequence:**
\ b (backspace BS, Unicode \u0008)
\ s (space SP, Unicode \u0020)
\ t (horizontal tab HT, Unicode \u0009)
\ n (linefeed LF, Unicode \u000a)
\ f (form feed FF, Unicode \u000c)
\ r (carriage return CR, Unicode \u000d)
\ [LineTerminator](https://docs.oracle.com/javase/specs/jls/se20/html/jls-3.html#jls-LineTerminator) (line continuation, no Unicode representation)
\ " (double quote ", Unicode \u0022)
\ ' (single quote ', Unicode \u0027)
\ \ (backslash \, Unicode \u005c)
[OctalEscape](https://docs.oracle.com/javase/specs/jls/se20/html/jls-3.html#jls-OctalEscape) (octal value, Unicode \u0000 to \u00ff)

**OctalEscape:**
\ [OctalDigit](https://docs.oracle.com/javase/specs/jls/se20/html/jls-3.html#jls-OctalDigit)
\ [OctalDigit](https://docs.oracle.com/javase/specs/jls/se20/html/jls-3.html#jls-OctalDigit) [OctalDigit](https://docs.oracle.com/javase/specs/jls/se20/html/jls-3.html#jls-OctalDigit)
\ [ZeroToThree](https://docs.oracle.com/javase/specs/jls/se20/html/jls-3.html#jls-ZeroToThree) [OctalDigit](https://docs.oracle.com/javase/specs/jls/se20/html/jls-3.html#jls-OctalDigit) [OctalDigit](https://docs.oracle.com/javase/specs/jls/se20/html/jls-3.html#jls-OctalDigit)

**OctalDigit:**
(one of)
0 1 2 3 4 5 6 7

**ZeroToThree:**
(one of)
0 1 2 3
